Refactored browser passcode authentication (Fixes #91)

kylehickinson opened the pull request:
After multiple chats with @jhreis we decided to refactor this code to better match the requirements of Brave's browser passcode and reduce bugs which may allow the user to bypass the lock.

Commit naming wise, doesn't prefix with "Fix X:" because it fixes it as a side-effect of refactor rather than the main purpose of this commit being fixing that bug.

## Pull Request Checklist

- [x] My patch has gone through review and I have addressed review comments
- [ ] My patch has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!`

- [ ] I have updated the *Unit Tests* to cover new or changed functionality
- [ ] I have updated the *UI Tests* to cover new or changed functionality
- [x] I have marked the bug with `[needsuplift]`
- [x] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

_None included_

## Notes for testing this patch

There are 4 permutations of the lock:
- **Requires Passcode Immediately**
- **Requires Passcode Immediately** and **Use Touch ID/Face ID**
- **Use Touch ID/Face ID**
- _(Both settings off)_

Commit:
Refactored browser passcode authentication (Fixes #91)

kylehickinson commented:
Sorry for label spam, thought there was a bug on device before realizing I wasn't on the right branch.

iccub commented:
We could tackle this https://github.com/brave/browser-ios/issues/1413 issue while we are at refactoring the lock screen.
Would be nice to prevent from showing popups if lock screen is present.

kylehickinson commented:
This may actually already be fixed due to the nature of using windows (or if not, may be easily fixed by changing the window level property), I'll run a test to see :)

kylehickinson commented:
Just tested and it looks like it worked in blocking alerts 👍

Locked | Unlocked
--- | ---
![blocked-alert](https://user-images.githubusercontent.com/529104/43080391-a40399ca-8e5d-11e8-8ecb-35f58b94b228.png) |  ![unlocked](https://user-images.githubusercontent.com/529104/43080400-a79889e2-8e5d-11e8-9424-e71b303420d7.png)

iccub commented:
Great stuff Kyle, I will comment in the v1 issue that we will get it fixed soon

jhreis commented on Client/Application/AppDelegate.swift:
```diff
@@ -190,6 +192,9 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
         window!.makeKeyAndVisible()
+        
+        authenticator = AppAuthenticator(protectedWindow: window!, isPasscodeEntryCancellable: false)
+        authenticator?.promptUserForAuthentication()
```
I think in most situations, these two lines of code will follow each other immediately. Probably cleanest to add a param to constructor `promptImmediately` or similar.

jhreis commented on Client/Application/AppDelegate.swift:
```diff
@@ -246,6 +251,8 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
     // We sync in the foreground only, to avoid the possibility of runaway resource usage.
     // Eventually we'll sync in response to notifications.
     func applicationDidBecomeActive(_ application: UIApplication) {
+        authenticator?.hideBackgroundedBlur()
```
Any timing issues between this and `applicationWillEnterForeground`?

kylehickinson replied:
The blurring happens on _will resign active_ and _did become active_ so we can't move it to _will enter foreground_ because there's no _will enter background_ delegate method (🤯) or else it wouldn't blur/unblur at the correct times

jhreis commented on Client/Application/AppDelegate.swift:
```diff
@@ -332,6 +339,16 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
+    }
+    
+    func applicationWillResignActive(_ application: UIApplication) {
+        if KeychainWrapper.sharedAppContainerKeychain.authenticationInfo() != nil {
```
Had to think about this, but yeah, I think blurring if any pincode option is set, is best.

jhreis commented on Client/Frontend/AuthenticationManager/AppAuthenticator.swift:
```diff
@@ -7,8 +7,108 @@ import Shared
 class AppAuthenticator {
-    static func presentAuthenticationUsingInfo(_ authenticationInfo: AuthenticationKeychainInfo, touchIDReason: String, success: (() -> Void)?, cancel: (() -> Void)?, fallback: (() -> Void)?) {
+    /// Whether or not its currently prompting the user to authenticate
+    private var isPrompting = false
```
This can be calculated via "window.isKeyWindow`. Probably better than attempting to manage ourselves.

kylehickinson replied:
Unfortunately, while I can make the show blur function use the `isKeyWindow` I can't get the hiding one to do it because of some race conditions happening with the touch ID/face ID prompt.

The state changes look like this basically:

Launch App → AppAuthenticator starts prompt, shows blur → (meanwhile: app becoming active, hide blur) → Prompt finally loads out of process and displays

The only thing we know is that when a biometric prompt comes up the app resigns active, and when it disappears the app becomes active again.  This isn't an issue with PIN entry because PIN entry is local.

jhreis commented on Client/Frontend/AuthenticationManager/AppAuthenticator.swift:
```diff
@@ -7,8 +7,108 @@ import Shared
+        UIView.animate(withDuration: 0.1, animations: { [weak window] in
+            window?.alpha = 0.0
+        }, completion: { [weak window] _ in
+            window?.isHidden = true
```
I think we should just `nil` out `window` here, no reason to keep it around. We can then just make `window` a lazy var so it is always valid. Setup `backgroundColor` and `rootViewController` inside there.

jhreis commented on Client/Frontend/AuthenticationManager/AppAuthenticator.swift:
```diff
@@ -7,8 +7,108 @@ import Shared
+        override func viewDidLoad() {
+            super.viewDidLoad()
+            
+            let backgroundedBlur = UIVisualEffectView(effect: UIBlurEffect(style: .light))
```
Can't we do something like:

```
UIVisualEffectView(effect: UIBlurEffect(style: .light)).then {
    view.addSubview(backgroundedBlur)
    backgroundedBlur.snp.makeConstraints { $0.edges.equalTo(self.view) }
}

kylehickinson replied:
Yeah we can, only thing is it'll actually end up being this: 

```swift
UIVisualEffectView(effect: UIBlurEffect(style: .light)).do {
    view.addSubview($0)
    $0.snp.makeConstraints { $0.edges.equalTo(self.view) }
}
```

The double `$0` may be a bit weird, but I can define the constraint maker explicitly like: `$0.snp.makeConstraints { make in make.edges.equalTo(self.view) }`

All-in-all, not sure if its really all that much better like using `Then.swift` in this case heh

jhreis commented on Client/Frontend/AuthenticationManager/AppAuthenticator.swift:
```diff
@@ -7,8 +7,108 @@ import Shared
+        }
+        
+        override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
+            return [.allButUpsideDown]
```
Why?

kylehickinson replied:
I thought it was needed due to the controller being on a separate window but it turns it out it's not, so… removed!

kylehickinson commented:
Updated based on feedback. One minor thing I also did: switched window alpha animations to use pop.  This way if you were to call `showBackgroundBlur` or `hideBackgroundBlur` any number of times it would reuse the animation rather than reset them and cause issues
